# VK API

VK API бесплатное, но нужно получить доступ


**Авторизация**

1. Перейти по ссылке https://vk.com/apps?act=manage
2. Создать новое приложение: standalone-приложение, название любое
3. Необходимо получить код подтверждения, может быть по номеру телефона
4. Если все ок, то вы перейдете в настройки приложения, там можно увидеть ключи доступа. На интересует **сервисный ключ доступа**

**Версии**

У API есть версии, поэтому вместе с токеном мы еще передаем информацию о том, какую версию мы используем. Версия отличает тем, какие параметры подаются, но больше это касается того, в каком виде мы получим ответ (структура ответа или наличие тех или иных полей может меняться от версии к версии).

In [3]:
import requests
from tqdm.auto import tqdm
from datetime import datetime
import time
import pandas as pd

Зафиксируем версию и токен в константах. В качестве токена подставляется сервисный ключ из настроек приложения. Укажем ту, которая дефолтно показывается в документации.

In [4]:
TOKEN = "5ddc69ee5ddc69ee5ddc69ee8f5eceded755ddc5ddc69ee3e2c0305f599d50fb0c3ca0c"
VERSION = "5.130"

## Метод wall.get

Первое, что мы сделаем - попробуем скачать информацию со стены, чтобы научиться собирать текстовые данные.

Подробная документация по ссылке: https://vk.com/dev/wall.get. Тут можно посмотреть на пример запроса, описание полей.

In [5]:
wall_get_url = "https://api.vk.com/method/wall.get" # endpoint, на который мы отправляем такие запросы

Запрашиваем 2 последних поста со страницы юзера с id = 1 (Павел Дуров). 

Для сообществ ID будут отрицательными (например, -1)

In [6]:
data = requests.get(
    wall_get_url, 
    params={
        "owner_id": 1,  # ID юзера
        "count": 2,  # кол-во постов
        "v": VERSION, # версия API
        "access_token": TOKEN  # токен доступа
    }
).json()

Мы получим ответ, который представляет собой словарь, где по ключу response лежит сам ответ.

Внутри лежит параметр count с числом записей (всего). В items сами посты (2, как мы просили).

Для каждого поста есть информация по объекту post. Подробное описание на странице https://vk.com/dev/objects/post

In [7]:
data

{'response': {'count': 1020,
  'items': [{'id': 2442097,
    'from_id': 1,
    'owner_id': 1,
    'date': 1525805964,
    'donut': {'is_donut': False},
    'comments': {'can_post': 0, 'count': 0, 'groups_can_post': True},
    'short_text_rate': 0.8,
    'attachments': [],
    'edited': 1525813826,
    'likes': {'can_like': 1,
     'count': 229448,
     'user_likes': 0,
     'can_publish': 1,
     'repost_disabled': False},
    'post_source': {'type': 'vk'},
    'post_type': 'post',
    'reposts': {'count': 5490, 'user_reposted': 0},
    'text': 'Иногда говорят, что Telegram был заблокирован в России, так как “закон есть закон”. Однако Telegram заблокирован в России как раз вопреки главному закону страны – Конституции. Решения судов и законы, противоречащие Конституции, не имеют силы. А это значит, что и сама блокировка Telegram незаконна. \n\nЕсли бы ФСБ ограничилась запросом информации о нескольких террористах, то ее требование вписывалось бы в рамки Конституции. Однако речь идет о пе

Можно заметить, что дата отображается в виде числа. Это специальный формат unixtimestamp, который очень часто используется, так как целые числа - это универсальный способ хранения, который можно исопльзовать в любой системе (JSON, любые БД и прочие)

In [6]:
unixtime = data['response']['items'][0]['date']
utc = datetime.fromtimestamp(unixtime)
print(unixtime, utc)

1525805964 2018-05-08 21:59:24


## wall.getComments

Следующий метод - это сбор комментариев к посту, что тоже может быть интересно для изучения диалогов или чего-то подобного.

Основные параметры: owner_id, post_id. Их мы можем достать из информации о постах на стене.

Можно это сделать и вручную, например, если открыть пост во всплывающем окне, то по адресу в адресной строке можно понять эти id.

https://vk.com/id1?w=wall1_2442097 : owner_id = 1, post_id = 2442097

Для примера возьмем пост из СМИ, где можно оставлять комментарии.

In [13]:
get_comments_url = "https://api.vk.com/method/wall.getComments"

In [16]:
data = requests.get(
    get_comments_url, 
    params={
        "owner_id": -76982440,
        "post_id": 5011073,
        "count": 2,
        "need_likes": 1,
        "v": VERSION,
        "access_token": TOKEN
    }
).json()

In [17]:
data

{'response': {'count': 16,
  'items': [{'id': 5011074,
    'from_id': 432646319,
    'post_id': 5011073,
    'owner_id': -76982440,
    'parents_stack': [],
    'date': 1618533695,
    'text': 'Страна советов',
    'likes': {'count': 14, 'user_likes': 0, 'can_like': 1},
    'thread': {'count': 0,
     'items': [],
     'can_post': True,
     'show_reply_button': True,
     'groups_can_post': True}},
   {'id': 5011078,
    'from_id': 174352380,
    'post_id': 5011073,
    'owner_id': -76982440,
    'parents_stack': [],
    'date': 1618533953,
    'text': 'Красавчик 👍🏼❤️',
    'likes': {'count': 0, 'user_likes': 0, 'can_like': 1},
    'thread': {'count': 0,
     'items': [],
     'can_post': True,
     'show_reply_button': True,
     'groups_can_post': True}}],
  'current_level_count': 10,
  'can_post': True,
  'show_reply_button': True,
  'groups_can_post': True}}

## groups.getMembers

Можно получить информацию о пользователях, например, о тех, кто подписан на опредленную группу.

Попробуем еще параметр offset, который может пригодиться, когда данных много и придется скачивать "страницы" или блоки по 100 или 1000 объектов (1000 - ограничение для этого метода)

In [7]:
group_members = "https://api.vk.com/method/groups.getMembers"

In [8]:
group = "dormitory8hse"

In [28]:
data = requests.get(
    group_members,
    params={
        'group_id': group,
        'access_token': TOKEN,
        'v': VERSION,
        'offset': 0
    }
).json()

Кол-во объектов всего

In [29]:
data["response"]["count"]

6456

В качестве ответа по людям - просто список ID. По ним уже дальше можно запрашивать подробную информацию о пользователях.

In [12]:
data["response"]["items"][:10]

[11952, 20090, 56613, 62028, 80420, 81206, 96206, 113393, 144980, 225392]

А теперь следующая страница. Это можно делать в цикле, чтобы выкачать всех

In [13]:
data = requests.get(
    group_members,
    params={
        'group_id': group,
        'access_token': TOKEN,
        'v': VERSION,
        'offset': 1000
    }
).json()

data["response"]["items"][:10]

[24723404,
 24729500,
 24745684,
 24943009,
 24952994,
 24966592,
 24979668,
 25002927,
 25011084,
 25017218]

## users.get

Попробуем собрать информацию о тех, кто попал

In [14]:
users_get_url = "https://api.vk.com/method/users.get"

In [20]:
import time

In [21]:
all_users_data = []

for user in tqdm(data["response"]["items"][:500]):
    time.sleep(0.2) # ОГРАНИЧЕНИЕ ПО ВРЕМЕНИ
    user_info = requests.get(
        users_get_url,
        params={
            'user_ids': user,
            'fields': 'bdate,home_town,universities',
            'access_token': TOKEN,
            'v': VERSION
        }
    ).json()
    all_users_data.extend(user_info["response"])

  0%|          | 0/500 [00:00<?, ?it/s]

In [16]:
user_info

{'error': {'error_code': 6,
  'error_msg': 'Too many requests per second',
  'request_params': [{'key': 'user_ids', 'value': '24966592'},
   {'key': 'fields', 'value': 'bdate,home_town,universities'},
   {'key': 'v', 'value': '5.130'},
   {'key': 'method', 'value': 'users.get'},
   {'key': 'oauth', 'value': '1'}]}}

In [25]:
df = pd.DataFrame(all_users_data)

# чтобы не показывать в открытом месте данные, возьмем для демонстрации толькодва столбца
df = df.dropna(subset=["home_town", "universities"])
df = df[["home_town", "universities"]] 

In [26]:
df.head()

home_town                                       universities
2   Тамбов, Гомель, Москва  [{'chair': 2037637, 'chair_name': 'Философия',...
3                           [{'city': 1, 'country': 1, 'education_form': '...
7              Владикавказ  [{'chair': 1767154, 'chair_name': 'Совместный ...
19                                                                         []
23                          [{'city': 1, 'country': 1, 'education_form': '...

Университеты нужно еще дальше разбирать, а вот города можно подсчитать сразу

In [27]:
df["home_town"].value_counts()

                             32
Уфа                           4
Омск                          4
Тверь                         4
Красноярск                    4
Москва                        3
Иркутск                       3
Ижевск                        2
Обнинск                       2
Курск                         2
Владикавказ                   2
Новороссийск                  1
Yerevan                       1
Киров                         1
Донецк                        1
Брянск                        1
Тамбов, Гомель, Москва        1
Тула                          1
Воронеж                       1
Ташкент                       1
С трамваями в метро...        1
Чайковский                    1
Удомля                        1
Тирасполь                     1
Нижний Новгород               1
Самара                        1
Новодвинск                    1
Гагино                        1
Оренбург                      1
Астана                        1
Асбест                        1
Абакан  

Видно, что не все идеально (есть сразу несколько), но после препроцессинга может быть ок.

In [30]:
# 

1000

## Задание

### 1

Попробовать собрать данные участников одной из групп и подсчитать статистику по какому-либо параметру (город, год рождения или иное).

1. Выбираем группу
2. Делаем пробный запрос
3. Пишем цикл по сбору (установка offset +1000 каждый раз)
4. Сохраняем ID
5. В цикле получаем информацию по пользователям
6. Сохраняем в датафрейм
7. Считаем статистику (и делаем препроцессинг, если нужно)

### 2

Собрать посты и/или комментарии, где авторы разные и попробовать сравнить употребление какого-либо слова у двух групп пользователей (Москва / не-Москва, Петербург / Москва, любые другие). Возможно, данных нужно будет много, но можно попробовать популярное слово. Главное - попробовать сделать основу для решения подобной задачи.

1. Выбираем группу
2. Собраем посты, сохраняем.
3. Для каждого поста собираем комментарии
4. По всем объектам собираем уникальные айди пользователей (лучше в порядке убывания частоты появления
5. Собираем информацию по ним, выделяем наши группы
6. Лемматизируем (если надо)
7. Считаем стастику